src/rules.js

```diff
@@ -8,8 +8,8 @@ const {
  * Block-Level Grammar
  */
 const block = {
-  newline: /^\n+/,
-  code: /^( {4}[^\n]+\n*)+/,
+  newline: /^(?: *(?:\n|$))+/,
+  code: /^( {4}[^\n]+(?:\n(?: *(?:\n|$))*)?)+/,
```

Comment:
```suggestion
  code: /^( {4}[^\n]+(?: *(?:\n|$))*)+/,
```

I *think* this simpler regex is equivalent.

Reply:
`[^\n]+(?: *)` is vulnerable to redos

Reply:
If that's the case, I'm happy to approve this PR. :thumbsup:

But I also really want to preserve simplicity as much as possible; could you first convince me of the redos issue? Maybe an example of a dangerous string for this regex? I'm failing to find anything that takes more than a few steps. Seems like `[^\n]+` would just greedily consume all characters including spaces up to the next `\n`, then we know `(?: *` does nothing 
and has no need to backtrack since the next character *must* be a `\n` (otherwise we would still be in `[^\n]+`).  And then we just loop on `(?: *(?:\n|$))*` which doesn't involve `[^\n]` at all.

Reply:
The problem is not with that regexp but with the fact that that regexp is imported into other regexps in rules.js

From https://www.regular-expressions.info/redos.html:

> A classic example is (.|\s)* to match any amount of any text when the regex flavor does not have a “dot matches line breaks” mode. If this is part of a longer regex then a subject string with a sufficiently long run of spaces will break the regex. The engine will try every possible combination of the spaces being matched by . or \s. For example, 3 spaces could be matched as ..., ..\s, .\s., .\s\s, \s.., \s.\s, \s\s., or \s\s\s. That’s 2^N permutations.

similarly here if the regex that this is part of fails even though `[^\n]+` is greedy if there is a space at the end it will try to put the space in the next group and try the regexp again.

Reply:
Gotcha. I didn't realize this was being included inside of other regexes.